File: Day13/Aditya239/solution.py
n, w, u, v, begin, end = [int(x) for x in raw_input().split()]
cross = float(w)/float(v)
u = float(u)
ships = []
K = 0.0
for i in range(n):
    laneinfo = raw_input().split()
    eastbound = 1 if laneinfo[0]=='E' else -1
    m = int(laneinfo[1])
    laneinfo = [float(k) for k in laneinfo[2:]]
    shipslane = [[-eastbound*laneinfo[2*j+1]/u - (i+1)*cross, (laneinfo[2*j] - eastbound*laneinfo[2*j+1])/u - i*cross] for j in range(m)]
    shipslane = [ship_ for ship_ in shipslane if ship_[0] < float(end)]
    ships += shipslane
ships=sorted(ships, key=lambda x: x[0])
free_since = float(begin)
for j in range(len(ships)):
    K = max(ships[j][0] - free_since, K)
    free_since = max(free_since, ships[j][1])
K = max(K, float(end) - free_since);
print(K)

File: Day13/Aditya239/solution.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
    int n, w, u, v, begin, end;
    cin>>n>>w>>u>>v>>begin>>end;
    vector<pair<double, double>> ships;
    double K = 0.0;
    vector<pair<int, int>> answer;
    char dir; int m; int nose; int len;
    double cross = double(w)/double(v);
    for(int j=0; j<n; j++){
        cin>>dir>>m;
        int eastbound = -1;
        if(dir=='E')
            eastbound = 1;
        for(int k=0; k<m; k++){
            cin >> len >> nose;
            double start = double(-eastbound*nose)/double(u) - (j+1)*cross;
            double stop  = double(len - eastbound*nose)/double(u) - j*cross;
            if(start>=double(end))
                continue;
            ships.push_back(pair<double, double>(start, stop));
        }
    }
    sort(ships.begin(), ships.end(), [](const std::pair<double, double> &left, const std::pair<double, double> &right){return left.first < right.first;});
    double free_since = double(begin);
    for(size_t j=0; j < ships.size(); j++){
        if(ships[j].first - free_since > K)
            K = ships[j].first - free_since;
        free_since = max(free_since, ships[j].second);
    }
    if(double(end) - free_since > K)
        K = double(end) - free_since;
    cout << K << endl;
    return 0;
}
